### PR TITLE
fix: pun-aliases for settings

### LINKED ISSUE #275: BaseSettings alias config not working
pydantic version `0.14`

```
❯ export test=foobar; pipenv run python
Python 3.6.1 (default, Aug  1 2018, 10:30:03)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> import pydantic
>>> class S(pydantic.BaseSettings):
...   test:str
...   class Config:
...     fields = dict(test=dict(alias="test"))
...
>>> S()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/jasonkuhrt/.local/share/virtualenvs/waldo-QOik3jd4/lib/python3.6/site-packages/pydantic/env_settings.py", line 35, in __init__
    super().__init__(**values)
  File "/Users/jasonkuhrt/.local/share/virtualenvs/waldo-QOik3jd4/lib/python3.6/site-packages/pydantic/main.py", line 166, in __init__
    self.__setstate__(self._process_values(data))
  File "/Users/jasonkuhrt/.local/share/virtualenvs/waldo-QOik3jd4/lib/python3.6/site-packages/pydantic/main.py", line 331, in _process_values
    return validate_model(self, input_data)
  File "/Users/jasonkuhrt/.local/share/virtualenvs/waldo-QOik3jd4/lib/python3.6/site-packages/pydantic/main.py", line 507, in validate_model
    raise ValidationError(errors)
pydantic.error_wrappers.ValidationError: 1 validation error
test
  none is not an allow value (type=type_error.none.not_allowed)
>>> os.getenv('test')
'foobar'
```

Expected that `test` would be parsed as str of `foobar`.

### PR DESCRIPTION
**Pull request by jasonkuhrt**
Will close #275

### CODE REVIEW

File: pydantic/fields.py [hunk 1]
```diff
@@ -69,6 +69,7 @@ def __init__(
 
         self.name: str = name
         self.alias: str = alias or name
+        self.pun_alias: bool = alias == name
```
Comment (jasonkuhrt) [inline pydantic/fields.py]: I felt adding a new field as the least intrusive solution.

The only real use-case for this field is to support settings which need to support pun-aliases to circumvent its default behaviour of:

- using var prefix
- using uppercase

Therefore maybe a comment is in order too, that explains such.

**Comment by samuelcolvin** [comment]
Why "pun"? Also I think we should respect alias regardless of whether it's the same as the field name even if it requires a bigger fix.

**Comment by jasonkuhrt** [comment]
Like how people came up with the term record puns. The alias is the same as thing thing it’s aliasing so it’s a “pun”.

What do you mean exactly by “respect the alias”?

**Comment by samuelcolvin** [comment]
I mean if you set the alias to `foobar` the env variable looked for should be `foobar` regardless of whether the field name happens to be `foobar` as well.

It might be more complicated to fix, but it'll make the library easier to understand and use.

**Comment by jasonkuhrt** [comment]
I believe that is the case now? We still check for alt_alias. And all the tests pass too. (They are failing on CI, don’t know why yet)

**Comment by samuelcolvin** [comment]
Failing on CI because of a linting error 

    tests/test_settings.py:111:9: E306 expected 1 blank line before a nested definition, found 0

> I believe that is the already the case with the PR as is?

Then we shouldn't need the `pun_alias` setting at all

**Comment by jasonkuhrt** [comment]
Not sure what you mean. If you remove my code change the test will fail for example. Are you suggesting you would like to see a solution that doesn’t require a new property on the fields class?

**Comment by samuelcolvin** [comment]
Yes exactly.

**Comment by jasonkuhrt** [comment]
@samuelcolvin gotchya. Agree the current solution is not ideal. Will see what I can come up with!

**Comment by jasonkuhrt** [comment]
@samuelcolvin started work on your feedback. Got stuck for the time being on:

```
❯ pytest -vv -x -k test_prevent_extra_success
Test session starts (platform: darwin, Python 3.6.5, pytest 3.8.1, pytest-sugar 0.9.1)
cachedir: .pytest_cache
rootdir: /Users/jasonkuhrt/projects/personal/pydantic, inifile: setup.cfg
plugins: sugar-0.9.1, mock-1.10.0, isort-0.2.1, cov-2.6.0, pylama-7.4.3


―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― test_prevent_extra_success ―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

    def test_prevent_extra_success():
        m = PreventExtraModel()
        assert m.foo == 'whatever'

>       m = PreventExtraModel(foo=1)

tests/test_main.py:170:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
pydantic/main.py:165: in __init__
    self.__setstate__(self._process_values(data))
pydantic/main.py:330: in _process_values
    return validate_model(self, input_data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

model = <[RecursionError("maximum recursion depth exceeded while calling a Python object") raised in repr()] PreventExtraModel object at 0x10c629848>, input_data = {'foo': 1}
```

**Comment by samuelcolvin** [comment]
It's not happening on master, must be caused by your change.

**Comment by jasonkuhrt** [comment]
@samuelcolvin yep. If you can shed any light feel free.

**Comment by samuelcolvin** [comment]
not sure I'm afraid, I'll try to look next week. Thanks for the work.

**Comment by samuelcolvin** [comment]
Once I started looking into this I decided you've over-complicated the solution (easy to do :smile:).

Was easier for me to replace this PR with #286

**Comment by jasonkuhrt** [comment]
@samuelcolvin Thanks for getting this through. I'm not sure what you mean though, since, your solution is literally the same as mine in 1346978cffe16ed95bb619ad76944664cd9382fe modulo superficial difference of `has_alias` ident name instead of `pun_alias`.

I only went a more complicated route upon your feedback that we should refactor the underlying way `alias` is treated.

I also stated most of this in https://github.com/samuelcolvin/pydantic/pull/282#discussion_r226812905 already.

🥂

**Pull request closed without merging**